larq_compute_engine/core/indirect_bgemm/kernel.h

```diff
@@ -1,123 +1,195 @@
-// parameter we choose: destination scalar; conv params; input/output shapes;
-// even detected CPU features.
-//     It is very important that this function is deterministic, as we rely on
-// the fact that the same kernel is selected for each call to `Eval` (as long as
-// the input shape doesn't change).
-template <typename DstScalar>
-inline IndirectBGEMMKernel<DstScalar> SelectRuntimeKernel(
-    const bconv2d::BConv2DParams* bconv2d_params,
-    const RuntimeShape& bitpacked_input_shape,
-    const RuntimeShape& output_shape) {
-#ifdef __aarch64__
-  // There are optimised assembly kernels for float and int8 output on Aarch64.
-  // They all use int16 accumulators. A different kernel is selected depending
-  // on whether the bitpacked number of input channels is a multiple of 4, 2, or
-  // 1, and whether that multiple is 1 or more than 1.
-
-  constexpr bool is_float_or_int8 = std::is_same<DstScalar, float>::value ||
-                                    std::is_same<DstScalar, std::int8_t>::value;
-  const int max_accumulator_value = bconv2d_params->filter_height *
-                                    bconv2d_params->filter_width *
-                                    bconv2d_params->channels_in;
-  const bool fits_in_int16_accumulators = max_accumulator_value + 512 < 1 << 16;
-
-  if (is_float_or_int8 && fits_in_int16_accumulators) {
-    // This weirdness is required because DstScalar could be `TBitpacked`, but
-    // in many cases RunKernel<TBitpacked, ...> isn't defined. As we can't use
-    // `if constexpr` (C++17), this causes a compile error despite the fact that
-    // we know that within this if block DstScalar must be float or int8.
-    using DS =
-        typename std::conditional<is_float_or_int8, DstScalar, float>::type;
-    using KFn = typename IndirectBGEMMKernel<DstScalar>::MicroKernelFunction;
-
-    if (bitpacked_input_shape.Dims(3) % 4 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 4) {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, true>, 8, 4, 4};
-      } else {
-        return {(KFn*)&kernel_8x4x4_aarch64::RunKernel<DS, false>, 8, 4, 4};
-      }
-    } else if (bitpacked_input_shape.Dims(3) % 2 == 0) {
-      if (bitpacked_input_shape.Dims(3) > 2) {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, true>, 8, 4, 2};
-      } else {
-        return {(KFn*)&kernel_8x4x2_aarch64::RunKernel<DS, false>, 8, 4, 2};
-      }
-    } else {
-      if (bitpacked_input_shape.Dims(3) > 1) {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, true>, 8, 4, 1};
-      } else {
-        return {(KFn*)&kernel_8x4x1_aarch64::RunKernel<DS, false>, 8, 4, 1};
+  const std::int32_t input_depth;
+  const std::int32_t output_channels;
+  const std::int32_t filter_size;
+  const std::int32_t groups;
+  const std::int32_t num_output_pixels;
+
+  std::vector<TBitpacked> packed_weights;
+  std::vector<const TBitpacked*> indirection_buffer;
+  std::vector<TBitpacked> zero_buffer;
```

Comment:
I've bundled everything related to the runtime kernel into a struct called `Kernel`, which stores all the information it needs (instead of needing these various vectors to be stored directly in the `op_data`, which was a bit messy).